akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/internal/EventOriginFilter.scala

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.projection.grpc.replication.internal
+
+import akka.annotation.InternalApi
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.typed.ReplicaId
+import akka.persistence.typed.internal.ReplicatedEventMetadata
+import akka.projection.grpc.internal.proto.InitReq
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class EventOriginFilter(selfReplicaId: ReplicaId) {
+
+  def createFilter(initReq: InitReq): EventEnvelope[_] => Boolean = {
+    // Events originating from other replicas that the consumer is connected to are excluded
+    // because the consumer will receive them directly from the other replica.
+    // Events originating from the consumer replica itself are excluded (break the cycle).
+    // Events originating from the producer replica are always included.
+    val exclude: Set[ReplicaId] =
+      initReq.otherReplicaIds.map(ReplicaId.apply).toSet ++
+      (if (initReq.replicaId == "") Nil else List(ReplicaId(initReq.replicaId))) -
+      selfReplicaId
+
+    { envelope =>
+      // Events from backtracking are lazily loaded via `loadEvent` if needed.
+      // Filter is done via `loadEvent` in that case.
+      envelope.eventMetadata match {
+        case Some(meta: ReplicatedEventMetadata) =>
+          !exclude(meta.originReplica)
+        case _ =>
+          // FIXME eventMetadata isn't loaded by backtracking
```

Comment:
It's working aside from this little detail. For ordinary transformation, which we used for the origin filtering, we solve it by not applying transformation for backtracking and instead do that in loadEvent. It would be difficult to do the same for this, because then we would need the replica ids also in the load event request.

Do you think it's fine to change the r2dbc query to always load metadata, also for backtracking?

Reply:
Sleeping on this. Feels fragile to not use the same origin filter from loadEvent also. Even though we only use loadEvent from backtracking events I think it should have the same transformation and filtering mechanism. I think it's best to include those additional replica id fields also in the loadEvent request.

> It would be difficult to do the same for this

no, that shouldn't be difficult

Reply:
I agree, let's go with always passing metadata and applying the same transform/filter.